wordpress: allow updating builtin themes and plugins

mohe2015 opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
To my knowledge it is currently not possible to update the themes and plugins that are shipped with wordpress by default.

**Also when migrating to the new method you fix the issue that you may lose theme/plugin settings on updates because the settings are stored based on the directory name and it changes as it contained the version.**

###### Things done
Split off https://github.com/NixOS/nixpkgs/pull/96910

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

r-rmcgibbo commented:
Result of `nixpkgs-review pr 135966` at 4d768ed8 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>wordpress-core</li>
  </ul>
</details>

mohe2015 commented:
Would love to get this into 21.11 if somebody wants to review this.

mohe2015 committed:
wordpress: Package default themes and plugins using wp4nix

mohe2015 committed:
nixos/wordpress: Migrate to wp4nix for themes and plugins and deprecate buggy option.

The original option used `name` instead of `pname` which may loose settings when upgrading themes or plugins.

onny commented:
I started to use the Wordpress module and now I can see the need for this. Having the standard themes and plugins in seperate packages is a good idea!

mohe2015 commented on pkgs/servers/web-apps/wordpress/default.nix:
```diff
@@ -12,17 +19,23 @@ stdenv.mkDerivation rec {
     cp -r . $out/share/wordpress
+    rm -r $out/share/wordpress/wp-content/plugins/*
+    rm -r $out/share/wordpress/wp-content/themes/*
+    ${lib.concatMapStringsSep "\n" (theme: "cp -r ${theme} $out/share/wordpress/wp-content/themes/${theme.passthru.wpName or theme.pname}") wpThemes}
```
It's been a (short) while but I think I'm currently breaking theme config for the built-in themes. maybe set wpThemes to null by default and only do these four lines here if it's not the default. Then users won't have the benefit of this when they're just using the wordpress package but we can at least hide this **really important** bugfix here without breaking existing users on accident. 

The bugfix is that the directory name doesn't contain the version as some themes/plugins are stupid enough to use that to find their config in the db.

mohe2015 replied:
The upstream package actually does this right so this should work.

Diff:
```
nix build --out-link original github:NixOS/nixpkgs#wordpress
nix build --out-link modified .#wordpress
diff --color -r original modified
```
looks good.

mohe2015 commented on nixos/modules/services/web-apps/wordpress.nix:
```diff
@@ -132,7 +153,27 @@ let
+                Then you can use the themes with e.g.
+                <code>
+                  (pkgs.callPackage "''${nixpkgs}/pkgs/servers/web-apps/wordpress/themes-and-plugins" {
+                    wp4nixDirectory = ../../../wordpress;
```
this too

mohe2015 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -32720,6 +32722,14 @@ with pkgs;
   wordpress = callPackage ../servers/web-apps/wordpress { };
 
+  wordpressPackages = callPackage ../servers/web-apps/wordpress/themes-and-plugins {
+    plugins = builtins.fromJSON (builtins.readFile ../servers/web-apps/wordpress/themes-and-plugins/plugins.json);
```
```suggestion
    plugins = lib.importJSON ../servers/web-apps/wordpress/themes-and-plugins/plugins.json;
```
...

mohe2015 commented:
This *should* be ready to review again. I changed the documentation a bit (and unfortunately duplicated some parts).

onny commented:
To simplify this PR I also created a separate one for including wp4nix https://github.com/NixOS/nixpkgs/pull/169183 

If this is merged we could update this one.

mohe2015 commented:
@GrahamcOfBorg test wordpress

onny commented on pkgs/servers/web-apps/wordpress/themes-and-plugins/default.nix:
```diff
@@ -0,0 +1,120 @@
+  } // lib.mapAttrs (type: pkgs: lib.makeExtensible (_: lib.mapAttrs (pname: data: self.mkOfficialWordpressDerivation { type = lib.removeSuffix "s" type; inherit pname data; }) pkgs)) generatedJson;
+
+# This creates an extensible scope and immediately extends it with our custom
+# package overrides.
```
Is this part required? If we want to keep it simple, it would be better to just include themes and plugins from the json files

onny replied:
Especially we could skip the visualcomposer and prismatic packages here?

mohe2015 replied:
Yeah I also thought about this. I like it though that this serves as an example on how to overrride this.

onny commented on pkgs/servers/web-apps/wordpress/update.sh:
```diff
@@ -0,0 +1,17 @@
+version=$(curl --globoff "https://api.wordpress.org/core/version-check/1.7/" | jq -r '.offers[0].version')
+update-source-version wordpress $version
+
+# cd /tmp
```
Are these comments still relevant?

mohe2015 replied:
You can use them to find out the officially packaged themes and plugins as they change at least every year with a newly packaged theme.

onny commented on pkgs/servers/web-apps/wordpress/update.sh:
```diff
@@ -0,0 +1,17 @@
+# ls wordpress/wp-content/plugins/ # update the command below with that
+
+cd pkgs/servers/web-apps/wordpress/themes-and-plugins
+WP_VERSION=5.8 WORKERS=1 wp4nix -t twentynineteen,twentytwenty,twentytwentyone -p akismet,hello-dolly
```
Is it correct that the version string is hardcoded here?

mohe2015 replied:
Not really but I I think you don't need to update it that often (I actually don't know exactly what it influences) and I think we can just update it with package updates. Inferring it from the .nix file is probably too complicated.

onny commented on pkgs/top-level/all-packages.nix:
```diff
@@ -30143,6 +30143,8 @@ with pkgs;
 
   wrapThunderbird = callPackage ../applications/networking/mailreaders/thunderbird/wrapper.nix { };
 
+  wp4nix = callPackage ../development/tools/wp4nix { };
```
wp4nix is already packaged and can be removed :)

onny commented on pkgs/top-level/all-packages.nix:
```diff
@@ -34491,6 +34493,14 @@ with pkgs;
+    themes = lib.importJSON ../servers/web-apps/wordpress/themes-and-plugins/themes.json;
+    languages = lib.importJSON ../servers/web-apps/wordpress/themes-and-plugins/languages.json;
+    pluginLanguages = lib.importJSON ../servers/web-apps/wordpress/themes-and-plugins/pluginLanguages.json;
+    themeLanguages = lib.importJSON ../servers/web-apps/wordpress/themes-and-plugins/themeLanguages.json;
```
language json files are missing in this PR. should we include the empty json files or is this safe when the file is missing?

mohe2015 replied:
I think I'll leave it. As nix is a lazy language it doesn't matter and it creates some visibility that this feature is still missing. Don't have any hard opinion on this though.

onny commented on pkgs/development/tools/wp4nix/default.nix:
```diff
@@ -0,0 +1,29 @@
+{ lib, buildGoModule, fetchFromGitLab, nix, subversion }:
```
This is already packaged and can be removed

onny commented on nixos/modules/services/web-apps/wordpress.nix:
```diff
@@ -298,6 +340,8 @@ in
           message = ''services.wordpress.sites."${hostName}".database.passwordFile cannot be specified if services.wordpress.sites."${hostName}".database.createLocally is set to true.'';
         }) eachSite);
 
+    warnings = mapAttrsToList (hostName: _: ''services.wordpress."${hostName}".plugins is deprecated and replaced by pkgs.wordpress.override { wpPlugins = [ ... ]; }. More info in the option documentation.'') (filterAttrs (n: v: v.plugins != []) eachSite)
```
I would really like to keep the plugins and themes option. Do you think using overrides is the better approach?

mohe2015 replied:
I don't necessarily think that using overrides is the better approach. This is mostly to make it possible to migrate themes and plugins from the $name-$version directory name to the $name directory name. This needs to be done as you loose settings on update for some themes+plugins with the old naming scheme. Before this I had a PR that did this based on `stateVersion` but it seemed like this would be too breaking as you couldn't migrate step by step. By doing it like this there should be no risk for breakage for existing packages.

mohe2015 replied:
I think another reason is that we need to remove the builtin themes to allow them to update and this can only be done easily in a backwards compatible way by doing this in the overridden package.

mweinelt replied:
Overriding the package can be done through the module system, which would make the options more discoverable.

mohe2015 commented:
A really unfortunate thing is that I'm currently not using NixOS any more so I won't be able to maintain this much. I'm still looking at NixOS stuff from time to time though.

mohe2015 commented on nixos/modules/services/web-apps/wordpress.nix:
```diff
@@ -82,7 +82,31 @@ let
           defaultText = literalExpression "pkgs.wordpress";
-          description = lib.mdDoc "Which WordPress package to use.";
+          example = literalExpression "pkgs.wordpress.override { wpPlugins = []; wpThemes = [ pkgs.wordpressPackages.themes.twentytwentyone ]; }";
+          description = ''
```
TODO: migrate to lib.mdDoc

mohe2015 commented:
It was decided out of band that this PR needs major rework (simplification). I will close this for now to avoid unecessary work for others reviewing etc.
